spinnaker_components/ensemble/recording.h

```diff
@@ -15,56 +15,97 @@
+  uint32_t *_sdram_start;    //!< Start of the buffer in SDRAM
+  uint32_t *_sdram_current;  //!< Current location in the SDRAM buffer
 } recording_buffer_t;
 
-/*!\brief Initialise a new recording buffer.
- */
-bool record_buffer_initialise(recording_buffer_t *buffer, address_t region,
-                              uint n_blocks, uint n_neurons);
-
 /*!\brief Reset the recording region for a new period of simulation.
  */
-void record_buffer_reset();
+void record_buffer_reset(recording_buffer_t *buffer);
 
 /*!\brief Flush the current buffer.
  *
  * The contents of the buffer will be appended to the recording region in
  * SDRAM, but only if recording is in use.
  */
-static inline void record_buffer_flush(recording_buffer_t *buffer) {
+static inline void record_buffer_flush(recording_buffer_t *buffer)
+{
   // Copy the current buffer into SDRAM
-  if (buffer->record) {
+  if (buffer->record)
+  {
     spin1_memcpy(buffer->_sdram_current, buffer->buffer,
                  buffer->block_length * sizeof(uint));
   }
 
   // Empty the buffer
-  for (uint i = 0; i < buffer->block_length; i++) {
-    buffer->buffer[i] = 0;
-  }
+  memset(buffer->buffer, 0x0, buffer->block_length * sizeof(uint32_t));
 
   // Progress the pointer
   buffer->_sdram_current += buffer->block_length;
 }
 
+/*****************************************************************************/
+/* Spike specific functions.
+ */
+
+/*!\brief Initialise a new recording buffer for recording spikes
+ */
+bool record_buffer_initialise_spikes(
+  recording_buffer_t *buffer,
+  address_t region,
+  uint n_blocks,
+  uint n_neurons
+);
+
 /*!\brief Record a spike for the given neuron.
  */
-static inline void record_spike(recording_buffer_t *buffer, uint n_neuron) {
+static inline void record_spike(recording_buffer_t *buffer, uint32_t n_neuron)
+{
   // Get the offset within the current buffer, and the specific bit to set
   // We write to the buffer regardless of whether recording is desired or not
   // in order to reduce branching.
   buffer->buffer[n_neuron >> 5] |= 1 << (n_neuron & 0x1f);
 }
 
+/*****************************************************************************/
+/* Voltage specific functions.
+ *
+ * As membrane potential in the normalised LIF implementation is clipped to [0,
+ * 1] we can discard the most significant 16 bits of the S16.15 representation
+ * to use U1.15 to represent the voltage without any loss.
+ */
+
+/*!\brief Initialise a new recording buffer for recording voltages
+ */
+bool record_buffer_initialise_voltages(
+  recording_buffer_t *buffer,
+  address_t region,
+  uint n_blocks,
+  uint n_neurons
+);
+
+/*!\brief Record a voltage for the given neuron.
+ */
+static inline void record_voltage(
+    recording_buffer_t *buffer, uint32_t n_neuron, value_t voltage
+)
+{
+  // Cast the recording buffer to U1.15 and then store the voltage in this
+  // form.
+  uint16_t *data = (uint16_t *) buffer->buffer;
+  data[n_neuron] = bitsk(voltage) & 0xffff;
```

Comment:
I'd contemplate using a union{ accum; uint16_t[2] } style construction here. Exactly what sign extending/casting occurs here is not immediately obvious